README.md
![scrcast](readme-header.png)

![Android CI](https://github.com/bmcreations/scrcast/workflows/Android%20CI/badge.svg)
![Latest Release](https://img.shields.io/github/v/release/bmcreations/scrcast)

A fully, featured replacement for screen recording needs backed by Kotlin with the power of Coroutines and Android Jetpack. scrcast is:

* <b>Easy to use</b>: scrcast's API leverages Kotlin languages features for simplicity, ease of use, and little-to-no boilerplate. Simply configure and `record()`
* <b>Modern</b>: scrcast is Kotlin-first and uses modern libraries including Coroutines and Android Jetpack.

## Download

scrcast is available on `mavenCentral()`.

`implementation ("dev.bmcreations:scrcast:$version")`

## Quick Start

scrcast provides a variety of configuration options for capturing, storing, and providing user interactions with your screen recordings.

### Configuring

```kotlin
val recorder = ScrCast.use(activity)
recorder.apply {
    // configure options via DSL
    options {
        video {
            maxLengthSecs = 360
        }
        storage {
            directoryName = "scrcast-sample"
        }
        notification {
            title = "Super cool library"
            description = "shh session in progress"
            icon = resources.getDrawable(R.drawable.ic_camcorder, null).toBitmap()
            channel {
                id = "1337"
                name = "Recording Service"
            }
            showStop = true
            showPause = true
            showTimer = true
        }
        moveTaskToBack = false
        startDelayMs = 5_000
    }
}
```

You can find [full configuration details and documentation here](https://bmcreations.github.io/scrcast/).

### State

interaction with `MediaRecorder`is abstracted in a easy to use and manage interface, via explict state-changing accessors.

#### Start

```kotlin
recorder.record()
```

#### Stop

```kotlin
recorder.stopRecording()
```

#### Pause

```kotlin
recorder.pause()
```

#### Resume

```kotlin
recorder.resume()
```

### Callbacks

State changes are emitted via `RecordingCallbacks` as a single interface or via a discrete lambda `onRecordingStateChange`

Completed recording output file is also emittable in `RecordingCallbacks` via

 ```kotlin
 fun onRecordingFinished(file: File)
 ```

## Requirements

* AndroidX
* `minSdkVersion` 23+
* `compileSdkVersion` 28+
* Java 8+

Gradle (`.gradle`)

```kotlin
android {
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
}

tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
    kotlinOptions {
        jvmTarget = "1.8"
    }
}
```

Gradle Kotlin DSL (`.gradle.kts`)

```kotlin
android {
    compileOptions {
        sourceCompatibility = JavaVersion.VERSION_1_8
        targetCompatibility = JavaVersion.VERSION_1_8
    }
}

tasks.withType<KotlinCompile> {
    kotlinOptions {
        jvmTarget = "1.8"
    }
}
```

## License

```text
Copyright 2020 bmcreations

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
#  0.3.0 (2021-01-12)

* **actions:** add dokka trigger for releases ([0072e4f](https://github.com/bmcreations/scrcast/commit/0072e4f5fa6315d181e056f88033963b3a166ac5))
* **readme:** add github action CI badge ([07d9e23](https://github.com/bmcreations/scrcast/commit/07d9e23031b7f50c590c845083b81b590dd4c2fd))
* chore: update mavan publish dependency; begin setup for action deploy ([526af18](https://github.com/bmcreations/scrcast/commit/526af18))

### Bug Fixes

* **scrcast:** fix media playback after recording; replace dispatcher with Activity Results API ([97788e9](https://github.com/bmcreations/scrcast/commit/97788e999c9816bf4000c4ddce564f373d176155)), closes [#46](https://github.com/bmcreations/scrcast/issues/46) [#48](https://github.com/bmcreations/scrcast/issues/48)

### Features

* **notifications:** Update default notification user experience ([#41](https://github.com/bmcreations/scrcast/issues/41)) ([15cc998](https://github.com/bmcreations/scrcast/commit/15cc998277a05670e63a8bfdb58c8690332e5206)), closes [#40](https://github.com/bmcreations/scrcast/issues/40)

## 0.2.0 (2020-07-24)

* feat: add support for maven deploy ([b1938c0](https://github.com/bmcreations/scrcast/commit/b1938c0))
* refactor(options/storage): have mediaStorageLocation self-generate ([b9f6909](https://github.com/bmcreations/scrcast/commit/b9f6909))
* refactor(scrcast): allow querying current options of instance ([745a35c](https://github.com/bmcreations/scrcast/commit/745a35c))
* docs(scrcast): add docs to exposed options ([b080054](https://github.com/bmcreations/scrcast/commit/b080054))
* chore(versioning): bump to 0.2.0 ([415cd4b](https://github.com/bmcreations/scrcast/commit/415cd4b))
* style: update README with logos ([e39b1b0](https://github.com/bmcreations/scrcast/commit/e39b1b0))
* fix(readme): update documentation link after transfer ([b65bb67](https://github.com/bmcreations/scrcast/commit/b65bb67))



## 0.1.0 (2020-07-19)

* chore: add docs generate and deploy script ([49faa12](https://github.com/bmcreations/scrcast/commit/49faa12))
* chore: add dokka to dependencies.gradle ([8a7ef45](https://github.com/bmcreations/scrcast/commit/8a7ef45))
* chore: add JVM sample with options, stage toggling ([877869e](https://github.com/bmcreations/scrcast/commit/877869e))
* chore: add maven-publish (#35) ([ac728dd](https://github.com/bmcreations/scrcast/commit/ac728dd)), closes [#35](https://github.com/bmcreations/scrcast/issues/35)
* chore: handle failures starting media recorder elegantly ([7ef6a65](https://github.com/bmcreations/scrcast/commit/7ef6a65))
* chore: remove changes to idea files ([2aae161](https://github.com/bmcreations/scrcast/commit/2aae161))
* chore: remove jitpack credential auth for now ([cba8661](https://github.com/bmcreations/scrcast/commit/cba8661))
* chore: remove maven plugin until we are public ([c878c32](https://github.com/bmcreations/scrcast/commit/c878c32))
* chore: remove tests ([1668466](https://github.com/bmcreations/scrcast/commit/1668466))
* chore: setup for jitpack ([217abea](https://github.com/bmcreations/scrcast/commit/217abea))
* chore: setup private jitpack ([a2e33cb](https://github.com/bmcreations/scrcast/commit/a2e33cb))
* chore: switch back to jitpack for now ([0710f26](https://github.com/bmcreations/scrcast/commit/0710f26))
* chore: upgrade gradle; reorganize receiver package ([6e5f9d8](https://github.com/bmcreations/scrcast/commit/6e5f9d8))
* chore(actions): swap install to build ([4393981](https://github.com/bmcreations/scrcast/commit/4393981))
* chore(actions): update build step & rm PR trigger ([96294b7](https://github.com/bmcreations/scrcast/commit/96294b7))
* chore(deps): unexpose coroutines ([bb84041](https://github.com/bmcreations/scrcast/commit/bb84041))
* chore(gradle): register gradle for IDE management ([bd7461b](https://github.com/bmcreations/scrcast/commit/bd7461b))
* chore(options): allow JVM overloads for config constructors ([9347810](https://github.com/bmcreations/scrcast/commit/9347810))
* chore(sample): add simple notification provider excert sample and init in app ([c98f136](https://github.com/bmcreations/scrcast/commit/c98f136))
* chore(sample): kill filewatcher from sample ([99a75c8](https://github.com/bmcreations/scrcast/commit/99a75c8))
* chore(sample): set output directory for sample ([4fbcd4c](https://github.com/bmcreations/scrcast/commit/4fbcd4c))
* fix: lazy load local broadcast manager ([e6b58a0](https://github.com/bmcreations/scrcast/commit/e6b58a0))
* fix(notification): hide pause actions when enabled if < API 24 ([5e3cf73](https://github.com/bmcreations/scrcast/commit/5e3cf73))
* fix(options): handle dynamic video sizing in scrcast instead of DSL ([5245fb9](https://github.com/bmcreations/scrcast/commit/5245fb9))
* fix(readme): correct gh pages linkage ([8b19456](https://github.com/bmcreations/scrcast/commit/8b19456))
* fix(recorder/service): create recorder _before_ creating virtual display from its surface ([98e0c96](https://github.com/bmcreations/scrcast/commit/98e0c96))
* fix(scrcast): allow directoryName and directory to be mutually exclusive ([019090d](https://github.com/bmcreations/scrcast/commit/019090d))
* fix(scrcast): don't scan for output file until state is reset to idle ([3a32a21](https://github.com/bmcreations/scrcast/commit/3a32a21))
* fix(scrcast/options): remove TS_2 conditional; increase bitrate to 8 Mbps ([5d5f8f7](https://github.com/bmcreations/scrcast/commit/5d5f8f7))
* fix(scrcast/recorder): start notification chronometer after start delay ([7914e9e](https://github.com/bmcreations/scrcast/commit/7914e9e))
* docs: Add getting started guide stub for scrcast ([445e15b](https://github.com/bmcreations/scrcast/commit/445e15b))
* docs: add initial README with TODO's (#33) ([7688b9a](https://github.com/bmcreations/scrcast/commit/7688b9a)), closes [#33](https://github.com/bmcreations/scrcast/issues/33)
* docs: finish README ([2e8c49f](https://github.com/bmcreations/scrcast/commit/2e8c49f))

app/src/main/java/dev/bmcreations/scrcast/app/list/MainActivity.kt
package dev.bmcreations.scrcast.app.list


import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.drawable.toBitmap
import androidx.core.view.isVisible
import com.google.android.material.snackbar.Snackbar
import dev.bmcreations.scrcast.ScrCast
import dev.bmcreations.scrcast.app.R
import dev.bmcreations.scrcast.extensions.supportsPauseResume
import dev.bmcreations.scrcast.recorder.RecordingState
import kotlinx.android.synthetic.main.activity_main.*


abstract class MainActivity : AppCompatActivity() {

    protected val recorder: ScrCast by lazy {
        ScrCast.use(this).apply {
            options {
                video {
                    maxLengthSecs = 360
                }
                storage {
                    directoryName = "scrcast-sample"
                }
                notification {
                    icon = resources.getDrawable(R.drawable.ic_camcorder, null).toBitmap()
                    channel {
                        id = "1337"
                        name = "Recording Service"
                    }
                    showStop = true
                    showPause = true
                    showTimer = true
                }
                moveTaskToBack = false
                stopOnScreenOff = true
                startDelayMs = 5_000
            }

            // ScrCast supports running your own notifications completely
            // simply provide a NotificationProvider
            //
            //setNotificationProvider(SimpleNotificationProvider(this@MainActivity))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        pause_fab.hide()
        pause_fab.setOnClickListener {
            if (recorder.state.isPaused) {
                recorder.resume()
            } else {
                recorder.pause()
            }
        }

        fab.setOnClickListener {
            if (recorder.state.isRecording) {
                recorder.stopRecording()
            } else {
                recorder.record()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        recorder.onRecordingComplete { file ->
            Snackbar.make(bottom_bar, "Recording located at ${file.absolutePath}", Snackbar.LENGTH_SHORT).show()
        }
    }

    override fun onStop() {
        super.onStop()
        recorder.onRecordingComplete { }
    }

    protected fun handleRecorderState(state: RecordingState) {
        Log.d("sample", "state change: state = $state")
        fab.reflectState(state)

        start_timer.isVisible = state.isInStartDelay
        if (supportsPauseResume) {
            if (state.isRecording || state.isPaused) {
                pause_fab.show()
            } else {
                pause_fab.hide()
            }
        }

        when (state) {
            is RecordingState.Delay -> start_timer.text = state.remainingSeconds.toString()
            RecordingState.Recording -> {
                pause_fab.setIconResource(R.drawable.ic_pause)
                pause_fab.text = "Pause"
            }
            is RecordingState.Idle -> {
                fab.isExpanded = false
                if (state.error != null) {
                    Toast.makeText(this, "Unable to start recording", Toast.LENGTH_SHORT).show()
                }
            }
            RecordingState.Paused -> {
                pause_fab.setIconResource(R.drawable.ic_resume)
                pause_fab.text = "Resume"
            }
        }
    }
}
